x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/ClusterDeprecationChecks.java

```diff
@@ -91,6 +95,53 @@ static DeprecationIssue checkTemplatesWithTooManyFields(ClusterState state) {
         return null;
     }
 
+    /**
+     * Check templates that use `_field_names` explicitly, which was deprecated in https://github.com/elastic/elasticsearch/pull/42854
+     * and will throw an error on new indices in 8.0
+     */
+    @SuppressWarnings("unchecked")
+    static DeprecationIssue checkTemplatesWithFieldNamesDisabled(ClusterState state) {
+        Set<String> templatesContainingFieldNames = new HashSet<>();
+        state.getMetaData().getTemplates().forEach((templateCursor) -> {
+            String templateName = templateCursor.key;
+            templateCursor.value.getMappings().forEach((mappingCursor) -> {
+                String type = mappingCursor.key;
+                // there should be the type name at this level, but there was a bug where mappings could be stored without a type (#45120)
+                // to make sure, we try to detect this like we try to do in MappingMetaData#sourceAsMap()
+                Map<String, Object> mapping = XContentHelper.convertToMap(mappingCursor.value.compressedReference(), true).v2();
+                if (mapping.size() == 1 && mapping.containsKey(type)) {
+                    // the type name is the root value, reduce it
+                    mapping = (Map<String, Object>) mapping.get(type);
+                }
+                if (mapContainsFieldNamesDisabled(mapping)) {
+                    templatesContainingFieldNames.add(templateName);
+                }
+            });
+        });
+
+        if (templatesContainingFieldNames.isEmpty() == false) {
+            return new DeprecationIssue(DeprecationIssue.Level.CRITICAL, "Index templates contain _field_names settings.",
```

Comment:
I think we should make this `WARNING`, as `CRITICAL` issues are generally reserved for things that will lead to loss of existing data or cause nodes to refuse to start if not addressed - they'll also prevent some external systems (e.g. ECE) from allowing upgrades until they're addressed.

While having index templates fail is bad, we've used `WARNING` for this type of issue in the past.

Reply:
Thanks for the comment, I used `CRITICAL` here for the template check because if those issues are not addressed, any logging use case with daily rotations (or beats for that matter) will not be able to continue after an upgrade, which could also lead to data loss (if index pipelines are overflowing etc...). Would you be absolutely oposed to this level? No problem changing this, just wanted to be rather safe than sorry.

Reply:
Also this should hopefully be relatively rare.

Reply:
The severity levels are defined in the docs at https://github.com/elastic/elasticsearch/blame/7.3/docs/reference/migration/apis/deprecation.asciidoc#L96-L100 and in a code comment at https://github.com/elastic/elasticsearch/blob/f3420afc8ead631dcfa523f1046e6f62922e8b26/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/deprecation/DeprecationIssue.java#L27-L36

It's definitely arguable that index templates that will cause index creation to fail falls under the technical definition of "will cause failures", it's not in line with how we have been using `CRITICAL`. We're also looking at tweaks to this API, including some changes which would allow us to cover situations like this better. Once we've moved on that, we'll review all the deprecation checks and change the severity level if necessary.